File: Java/DataStruct_cirSingle.java
public class DataStruct_cirSingle {
    public static void main(String[] args){
        cirLinkList list1=new cirLinkList();
        for(int i=0;i<=5;i++){
            list1.addFirst(i);
        }
        list1.sortMinMax();
        list1.show();
        
    }
}
class node{
    int num;
    node next;

    node(int numIn){
        num=numIn;
        next=null;
    }
}
class cirLinkList{
    int amount;
    int currPos;
    node nullNode=new node(-1);
    node head;
    node tail;

    cirLinkList(){
        currPos=1;
        amount=0;
        head=null;
        tail=null;
    }
    void addFirst(int newNum){
        node newNode=new node(newNum);
        if((head==null)&&(tail==null)){
            head=newNode;
            tail=newNode;
        }
        else{
            newNode.next=head;
            head=newNode;
        }
        tail.next=head;
        amount++;
    }
    void addLast(int newNum){
        node newNode=new node(newNum);
        if((head==null)&&(tail==null)){
            head=newNode;
        }
        else{
            newNode.next=head;
            tail.next=newNode;
        }
        tail=newNode;
        amount++;
    }
    void show(){
        node currNode=head;
        currPos=1;
        while(currPos<=amount){
            System.out.printf("%d ", currNode.num);
            currNode=currNode.next;
            currPos++;
        }
        System.out.printf("\n\n");
    }
    node findNode(node currNode, int currPos, int pos){
        if(currPos>amount){
            return nullNode;
        }
        else{
            if(currPos==pos){
                return currNode;
            }
            else{
                return findNode(currNode.next, currPos+1, pos);
            }
        }
    }
    node findNum(int num){
        node currNode=head;
        currPos=1;
        while(currPos<=amount){
            if(currNode.num==num){
                return currNode;
            }
            else{
                currNode=currNode.next;
                currPos++;
            }
        }
        return nullNode;
    }
    node findPrevNum(int num){
        node currNode=findNum(num);
        node prevNode=head;
        currPos=1;
        while(currPos<=amount){
            if(prevNode.next==currNode){
                return prevNode;
            }
            else{
                prevNode=prevNode.next;
                currPos++;
            }
        }
        return nullNode;
    }
    node findPrevNode(node currNode, int currPos, int pos){
        if(pos>amount){
            return tail;
        }
        else if(pos<=1){
            return head;
        }
        else{
            if(currPos==pos-1){
                return currNode;
            }
            else{
                return findPrevNode(currNode.next, currPos+1, pos);
            }
        }
    }
    void deleteFirst(){
        node temp=head;
        head=head.next;
        tail.next=head;
        temp.next=null;
        amount--;
    }
    void deleteLast(){
        node temp=tail;
        tail=findNode(head, 1, amount-1);
        tail.next=head;
        temp.next=null;
        amount--;
    }
    void deleteAt(int pos){
        if(pos<=1){
            deleteFirst();
        }
        else if(pos>=amount){
            deleteLast();
        }
        else{
            node currNode=findNode(head, 1, pos);
            node prevNode=findNode(head, 1, pos-1);
            prevNode.next=currNode.next;
            currNode.next=null;
            amount--;
        }
    }
    void insertAt(int pos, int newNum){
        node newNode=new node(newNum);
        if(pos<=1){
            addFirst(newNum);
        }
        else if(pos>amount){
            addLast(newNum);
        }
        else{
            node prevNode=findNode(head, 1, pos-1);
            node currNode=findNode(head, 1, pos);
            newNode.next=currNode;
            prevNode.next=newNode;
            amount++;
        }
    }
    void swapNode(int pos1, int pos2){
        if(pos1<1){
            pos1=1;
        }
        else if(pos1>amount){
            pos1=amount;
        }
        if(pos2<1){
            pos2=1;
        }
        else if(pos2>amount){
            pos2=amount;
        }
        node node1=findNode(head, 1, pos1);
        node node2=findNode(head, 1, pos2);
        if((node1!=null)&&(node2!=null)){
            int temp=node1.num;
            node1.num=node2.num;
            node2.num=temp;
        }
        else{
            System.out.println("Can't swap.");
        }
    }
    void swapNum(int num1, int num2){
        node node1=findNum(num1);
        node node2=findNum(num2);
        if((node1!=nullNode)&&(node2!=nullNode)){
            int temp=node1.num;
            node1.num=node2.num;
            node2.num=temp;
        }
        else{
            System.out.println("Can't swap.");
        }
    }
    void sortMinMax(){
        node currNode=head;
        node lastNode=tail;
        while(lastNode!=head){
            while(currNode!=lastNode){
                if(currNode.num>currNode.next.num){
                    swapNum(currNode.num, currNode.next.num);
                }
                currNode=currNode.next;
            }
            currNode=head;
            lastNode=findPrevNum(lastNode.num);
        }
    }
}

File: C++/Data_Structure/LinkList/DataStruct_cirSingle.cpp
#include<iostream>
using namespace std;
class node{
    public:
        int num;
        node *next;

        node(int newNum){
            num=newNum;
            next=NULL;
        }
};
class cirLinkList{
    public:
        int amount;
        node *nullNode=new node(-1);
        node *head;
        node *tail;

        cirLinkList(){
            amount=0;
            head=NULL;
            tail=NULL;
        }
        void addFirst(int newNum);
        void addLast(int newNum);
        void show();
        node *findNode(node *currNode, int walk, int pos);
        node *findNum(node *currNode, int target);
        void deleteFirst();
        void deleteLast();
        void insertAt(int pos, int newNum);
        void deleteAt(int pos);
};
void cirLinkList::addFirst(int newNum){
    node *newNode=new node(newNum);
    if((head==NULL)&&(tail==NULL)){
        tail=newNode;
    }
    else{
        newNode->next=head;
        tail->next=newNode;
    }
    head=newNode;
    amount++;
}
void cirLinkList::addLast(int newNum){
    node *newNode=new node(newNum);
    if((head==NULL)&&(tail==NULL)){
        head=newNode;
    }
    else{
        newNode->next=head;
        tail->next=newNode;
    }
    tail=newNode;
    tail->next=head;
    amount++;
}
void cirLinkList::show(){
    node *currNode=head;
    while(currNode!=tail){
        cout<<currNode->num<<" ";
        currNode=currNode->next;
    }
    cout<<tail->num<<endl<<endl;
}
node *cirLinkList::findNode(node *currNode, int walk, int pos){
    if(walk==amount){
        return nullNode;
    }
    else{
        if(walk==pos-1){
            return currNode;
        }
        else{
            return findNode(currNode->next, walk+1, pos);
        }
    }
}
node *cirLinkList::findNum(node *currNode, int target){
    if((currNode==tail)&&(currNode->num!=target)){
        return nullNode;
    }
    else{
        if(currNode->num==target){
            return currNode;
        }
        else{
            return findNum(currNode->next, target);
        }
    }
}
void cirLinkList::deleteFirst(){
    node *temp=head;
    head=temp->next;
    tail->next=head;
    temp->next=NULL;
    amount--;
}
void cirLinkList::deleteLast(){
    node *temp=tail;
    tail=findNode(head, 0, amount-1);
    tail->next=head;
    temp->next=NULL;
    amount--;
}
void cirLinkList::insertAt(int pos, int newNum){
    node *currNode=findNode(head, 0, pos);
    if(currNode==head){
        addFirst(newNum);
    }
    else if(currNode==tail){
        addLast(newNum);
    }
    else{
        node *newNode=new node(newNum);
        node *prevNode=findNode(head, 0, pos-1);
        newNode->next=currNode;
        prevNode->next=newNode;
    }
    amount++;
}
void cirLinkList::deleteAt(int pos){
    node *targetNode=findNode(head, 0, pos);
    if(targetNode==head){
        deleteFirst();
    }
    else if(targetNode==tail){
        deleteLast();
    }
}
int main(){
    cirLinkList cirLL;
    for(int i=0;i<=5;i++){
        cirLL.addFirst(i);
    }
    cirLL.show();
    cout<<cirLL.findNum(cirLL.head, 3)->num;
}